timemachine/fe/atom_mapping.py

```diff
@@ -1,290 +1,435 @@
+            complete_rings,
+        )
+        return all_cores
+
+
+def bfs(g, atom):
+    depth = 0
+    cur_layer = [atom]
+    levels = {}
+    while len(levels) != g.GetNumAtoms():
+        next_layer = []
+        for layer_atom in cur_layer:
+            levels[layer_atom.GetIdx()] = depth
+            for nb_atom in layer_atom.GetNeighbors():
+                if nb_atom.GetIdx() not in levels:
+                    next_layer.append(nb_atom)
+        cur_layer = next_layer
+        depth += 1
+    levels_array = [-1] * g.GetNumAtoms()
+    for i, l in levels.items():
+        levels_array[i] = l
+    return levels_array
+
+
+def reorder_atoms_by_degree(mol):
+    degrees = [len(a.GetNeighbors()) for a in mol.GetAtoms()]
+    perm = np.argsort(degrees)[::-1]
+    new_mol = Chem.RenumberAtoms(mol, perm.tolist())
+    return new_mol, perm
+
+
+def find_cycles(g: nx.Graph):
+    # return the indices of nxg that are in a cycle
+    # 1) find and remove bridges
+    # 2) if atom has > 1 neighbor then it is in a cycle.
+    edges = nx.bridges(g)
+    for e in edges:
+        g.remove_edge(*e)
+    cycle_dict = {}
+    for node in g.nodes:
+        # print(list(g[node]))
+        cycle_dict[node] = len(list(g.neighbors(node))) > 0
+    return cycle_dict
+
+
+def get_edges(mol):
+    return [(bond.GetBeginAtomIdx(), bond.GetEndAtomIdx()) for bond in mol.GetBonds()]
+
+
+def induce_subgraphs_given_marcs(mol_a, mol_b, marcs):
+    sg_a = nx.Graph()
+    sg_b = nx.Graph()
+    edges_a = get_edges(mol_a)
+    edges_b = get_edges(mol_b)
+    for e1 in range(marcs.shape[0]):
+        src_a, dst_a = edges_a[e1]
+        for e2 in range(marcs.shape[1]):
+            src_b, dst_b = edges_b[e2]
+            if marcs[e1][e2]:
+                sg_a.add_edge(src_a, dst_a)
+                sg_b.add_edge(src_b, dst_b)
+
+    return sg_a, sg_b
+
+
+def _to_networkx_graph(mol):
+    g = nx.Graph()
+    for atom in mol.GetAtoms():
+        g.add_node(atom.GetIdx())
+
+    for bond in mol.GetBonds():
+        src, dst = bond.GetBeginAtomIdx(), bond.GetEndAtomIdx()
+        g.add_edge(src, dst)
+    return g
+
+
+def _remove_incomplete_rings(mol_a, mol_b, core, marcs):
+
+    # to networkx
+    g_a, g_b = _to_networkx_graph(mol_a), _to_networkx_graph(mol_b)
+    sg_a, sg_b = induce_subgraphs_given_marcs(mol_a, mol_b, marcs)
+
+    a_cycles = find_cycles(g_a)
+    b_cycles = find_cycles(g_b)
+    sg_a_cycles = find_cycles(sg_a)
+    sg_b_cycles = find_cycles(sg_b)
+
+    new_core = []
+    new_marcs = marcs.copy()
+
+    mcg_a = mcgregor.Graph(mol_a.GetNumAtoms(), get_edges(mol_a))
+    mcg_b = mcgregor.Graph(mol_b.GetNumAtoms(), get_edges(mol_b))
+
+    for a, b in core:
+        # subgraph and graph are consistent
+        pred_sgg_a = a_cycles[a] == sg_a_cycles[a]
+        pred_sgg_b = b_cycles[b] == sg_b_cycles[b]
+        # a and b are consistent
+        pred_sg_ab = sg_a_cycles[a] == sg_b_cycles[b]
+        pred_g_ab = a_cycles[a] == b_cycles[b]
```

Comment:
@jkausrelay invariance for complete rings:

We require a strong consistency that each mol's subgraph and graph have the same ring requirement, and that mol_a's subgraph is consistent with mol_b's subgraph. I think the last assertion that mol_a's graph is consistent with mol_b's graph may be redundant (and would be enforced by atom-atom predicates, which turns out to be unnecessary)

Reply:
Makes sense to me